clients/drcachesim/tracer/output.cpp

```diff
@@ -1365,21 +1381,19 @@ exit_thread_io(void *drcontext)
     // Append a thread exit marker and output remaining records for this thread
     // if we are still in a tracing mode or the thread has data from a prior window
     // that it never wrote out.
-    if (is_in_tracing_mode(tracing_mode.load(std::memory_order_acquire)) ||
-        (has_tracing_windows() &&
+    if ((has_tracing_windows() &&
          // If non-split we always want to append a thread exit marker as there
          // wouldn't be on otherwise (split has one at the end of each window file).
          (!op_split_windows.get_value() ||
           // If split, we only need to write if we have data from a prior window.
           (get_local_window(data) < tracing_window.load(std::memory_order_acquire) &&
            !is_new_window_buffer_empty(data)))) ||
-        // For attach we switch to BBDUP_MODE_NOP but still need to finalize
-        // each thread.  However, we omit threads that did nothing the entire time
-        // we were attached.
-        (!has_tracing_windows() && align_attach_detach_endpoints() &&
-         (data->bytes_written > 0 ||
-          BUF_PTR(data->seg_base) - data->buf_base >
-              static_cast<ssize_t>(data->init_header_size + buf_hdr_slots_size)))) {
+        // Throw out empty threads even in a tracing mode.
+        ((data->bytes_written > 0 || buffer_contains_nontrivial_data(data)) &&
```

Comment:
Curious why we need both data->bytes_written > 0 and buffer_contains_nontrivial_data(data). The latter seems a stricter condition; does it happen that one is true without the other?

Reply:
They are quite different: bytes_written tells us whether prior buffer contents have already been written out (if so, we definitely want to close out this thread with an exit record), while the other check is about the just the current buffer.  We only want to ignore a thread that both has a trivially-empty current buffer and never wrote anything before.

Added a comment.